src/app/qgisapp.cpp

```diff
@@ -5283,34 +5295,47 @@ bool QgisApp::fileSave()
   // that the project file name is reset to null in fileNew()
   QFileInfo fullPath;
 
-  if ( QgsProject::instance()->fileName().isNull() )
+  if ( QgsProject::instance()->fileName().isNull() && QgsProject::instance()->zipFileName().isNull() )
   {
     // Retrieve last used project dir from persistent settings
     QgsSettings settings;
     QString lastUsedDir = settings.value( QStringLiteral( "UI/lastProjectDir" ), QDir::homePath() ).toString();
 
+    const QString qgs_ext = tr( "QGIS files" ) + " (*.qgs)";
+    const QString zip_ext = tr( "QGZ files" ) + " (*.qgz)";
+    QString filter;
     QString path = QFileDialog::getSaveFileName(
                      this,
                      tr( "Choose a QGIS project file" ),
                      lastUsedDir + '/' + QgsProject::instance()->title(),
-                     tr( "QGIS files" ) + " (*.qgs *.QGS)" );
+                     qgs_ext + ";;" + zip_ext, &filter );
     if ( path.isEmpty() )
       return false;
 
     fullPath.setFile( path );
 
     // make sure we have the .qgs extension in the file name
-    if ( "qgs" != fullPath.suffix().toLower() )
+    if ( filter == zip_ext )
     {
-      fullPath.setFile( fullPath.filePath() + ".qgs" );
-    }
+      if ( "qgz" != fullPath.suffix().toLower() )
```

Comment:
qt optimisation tip: toLower() for case insensitive matching is inefficient (requires that the whole string be converted to lowercase, the new lowercase string needs memory allocation, etc). Use .compare( ... , Qt::CaseInsensitive) instead, as that only needs to process as many characters in the original QString objects as are required before a mismatch is found.